core/src/storage2/collections/stash/mod.rs

```diff
@@ -0,0 +1,549 @@
+    ///
+    /// # Note
+    ///
+    /// The `removed_index` points to the index of the removed vacant entry.
+    fn remove_vacant_entry(&mut self, removed_index: Index, vacant_entry: VacantEntry) {
+        let prev_vacant = vacant_entry.prev;
+        let next_vacant = vacant_entry.next;
+        if prev_vacant == removed_index && next_vacant == removed_index {
+            // There is no other vacant entry left in the storage stash so
+            // there is nothing to update. Bail out early.
+            return
+        }
+        if prev_vacant == next_vacant {
+            // There is only one other vacant entry left.
+            // We can update the single vacant entry in a single look-up.
+            if let Some(entry) = self
+                .entries
+                .get_mut(prev_vacant)
+                .map(Entry::try_to_vacant_mut)
+                .expect("`prev` must point to an existing entry at this point")
+            {
+                debug_assert_eq!(entry.prev, removed_index);
+                debug_assert_eq!(entry.next, removed_index);
+                entry.prev = prev_vacant;
+                entry.next = prev_vacant;
+            }
+        } else {
+            // There are multiple other vacant entries left.
+            if let Some(entry) = self
+                .entries
+                .get_mut(prev_vacant)
+                .map(Entry::try_to_vacant_mut)
+                .expect("`prev` must point to an existing entry at this point")
+            {
+                debug_assert_eq!(entry.next, removed_index);
+                entry.next = next_vacant;
+            }
+            if let Some(entry) = self
+                .entries
+                .get_mut(next_vacant)
+                .map(Entry::try_to_vacant_mut)
+                .expect("`next` must point to an existing entry at this point")
+            {
+                debug_assert_eq!(entry.prev, removed_index);
+                entry.prev = prev_vacant;
+            }
+        }
+        // Bind the last vacant pointer to the vacant position with the lower index.
+        // This has the effect that lower indices are refilled more quickly.
+        use core::cmp::min;
+        if removed_index == self.header.last_vacant {
+            self.header.last_vacant = min(prev_vacant, next_vacant);
+        }
+    }
+
+    /// Put the element into the stash at the next vacant position.
+    ///
+    /// Returns the stash index that the element was put into.
+    pub fn put(&mut self, new_value: T) -> Index {
+        let new_entry = Some(Entry::Occupied(new_value));
+        let new_index = if let Some(index) = self.last_vacant_index() {
+            // Put the new element to the most recent vacant index if not all entries are occupied.
+            let old_entry = self
+                .entries
+                .put_get(index, new_entry)
+                .expect("a `next_vacant` index must point to an occupied cell");
```

Comment:
expect message appears out of sync. Perhaps you mean `last_vacant_index()` rather than `next_vacant`